### PR TITLE
Add init to Provider contract

### PR DESCRIPTION
**Pull request by woop**
**What this PR does / why we need it**:

This PR adds an abstract `__init__` as part of the base class for Providers. Currently no such method exists which has led to custom providers and built in providers having different initialization contracts.
https://github.com/feast-dev/feast/blob/745a1b43d20c0169b675b1f28039854205fb8180/sdk/python/feast/infra/provider.py#L146
vs 
https://github.com/feast-dev/feast/blob/745a1b43d20c0169b675b1f28039854205fb8180/sdk/python/feast/infra/provider.py#L164

**Does this PR introduce a user-facing change?**:

```release-note
All custom providers must now provide an initialization method and won't receive a repo_path argument
```

### CODE REVIEW

File: sdk/python/feast/infra/provider.py [hunk 1]
```diff
@@ -20,6 +20,10 @@
 
 
 class Provider(abc.ABC):
+    @abc.abstractmethod
+    def __init__(self, config: RepoConfig):
```
Comment (achals) [inline sdk/python/feast/infra/provider.py:24]: Won't we break existing third party providers?

Comment (woop) [reply sdk/python/feast/infra/provider.py:24]: Yes.

Comment (achals) [reply sdk/python/feast/infra/provider.py:24]: Should we also update the call site to not pass in repo_path when loading custom providers?

Comment (woop) [reply sdk/python/feast/infra/provider.py:24]: Yea, good catch.

**Pull request merged**